### PR TITLE
Per role config with status

### PR DESCRIPTION
**Pull request by ilackarms**
config is separated by role for each virtual service

psuedo code reflecting changes:
```
foreach vs : cfg.virtualservices:
  foreach role: vs.roles:
    map[role] = append(map[role], vs)

foreach role, []vss : map
  create_role(role)
  filtered_upstreams = filter_upstreams(allUpstreams, []vss)
  snap, reports = translate(role, []vss, filtered_upstreams)
  xds.set_snapshot(role, snap)
  write_reports(reports)
```

writing a report for the role will create the role if it doesnt exist; this is because roles are currently read-only

### CODE REVIEW

File: test/kube_e2e/kubernetes_suite_test.go [hunk 1]
```diff
@@ -17,8 +19,10 @@ func TestKubernetes(t *testing.T) {
 
 	helpers.RegisterPreFailHandler(func() {
 		var logs string
+		limitLogLines := 250
 		for _, component := range []string{"control-plane", "ingress"} {
 			l, err := helpers.KubectlOut("logs", "-l", "gloo="+component)
+			l = strings.Join(strings.Split(l, "\n")[:limitLogLines], "\n")
```
Comment (yuval-k) [inline test/kube_e2e/kubernetes_suite_test.go]: Should we show log lines from the end or the beginning?

Comment (ilackarms) [reply test/kube_e2e/kubernetes_suite_test.go]: probably the end

File: internal/control-plane/eventloop/eventloop.go [hunk 2]
```diff
@@ -135,25 +138,141 @@ func (e *eventLoop) updateXds(snap *snapshot.Cache) {
...
+		}
+	}
 
-	xdsSnapshot, reports, err := e.translator.Translate(snap)
-	if err != nil {
-		// TODO: panic or handle these internal errors smartly
-		log.Warnf("failed to translate based on the latest config: %v", err)
-		return
+	// aggregate reports across all the roles
+	allReports := make(map[string]reporter.ConfigObjectReport)
```
Comment (yuval-k) [inline internal/control-plane/eventloop/eventloop.go]: why is this a map and not an array?

File: api/v1/role.proto [hunk 3]
```diff
@@ -16,19 +16,22 @@ import "metadata.proto";
...
+    // Currently this is done in the format <Role>~<this portion is ignored>
     // which can be specified with the `--service-node` flag, or in the Envoy instance's bootstrap config.
     //
-    // Names must be unique and follow the following syntax rules:
+    // Role Names must be unique and follow the following syntax rules:
     // One or more lowercase rfc1035/rfc1123 labels separated by '.' with a maximum length of 253 characters.
     string name = 1;
 
-    // the list of names of the virtual services this role encapsulates.
+    // a list of virtual services that reference this role
```
Comment (yuval-k) [inline api/v1/role.proto]: This field is read only right? i would mention that in the comment

File: internal/control-plane/eventloop/eventloop.go (hunk 2 above)
Comment (ilackarms) [reply internal/control-plane/eventloop/eventloop.go]: some reports get re-generated by the translation loop; any resource (upstream or virtualservice) that appears for more than one role will get regenerated. this is to de-duplicate reports for the same object

File: api/v1/role.proto (hunk 3 above)
Comment (ilackarms) [reply api/v1/role.proto]: the whole object is read-only, currently. there's a comment at the top of the file for this

**Pull request merged**